### PR TITLE
feat: add left join

### PR DESCRIPTION
**Pull request by mfonekpo**
## What type of PR is this? (check all applicable)

- [ ] 💾 Refactor
- [x] ✨ Feature
- [ ] 🐛 Bug Fix
- [ ] 🔧 Optimization
- [ ] 📝 Documentation
- [x] ✅ Test
- [ ] 🐳 Other

## Related issues 

- Related issue # 
- Closes #

## Checklist

- [ ] Code follows style guide (ruff)
- [x] Tests added 
- [ ] Documented the changes

### CODE REVIEW

File: narwhals/dataframe.py [hunk 1]
```diff
@@ -181,6 +181,23 @@ def join(
             )
         )
 
+    def left_join(
```
Comment (MarcoGorelli) [inline narwhals/dataframe.py:184]: i think we want `join(how='left')`, not `left_join`?

File: tests/test_common.py [hunk 2]
```diff
@@ -271,8 +271,10 @@ def test_join(df_raw: Any) -> None:
     expected = {"a": [1, 3, 2], "b": [4, 4, 6], "z": [7.0, 8, 9], "z_right": [7.0, 8, 9]}
     compare_dicts(result_native, expected)
 
+    result = df.join(df_right, left_on="a", right_on="a", how="left")  # type: ignore[arg-type]
```
Comment (MarcoGorelli) [inline tests/test_common.py:274]: cool, right looks right - now, you need to construct `expected`, and then check that it matches `result` using `compare_dicts` (see line 272)

**Review by MarcoGorelli — COMMENTED** [review]
Nice one @mfonekpo - getting closer!

Something we need to take care of is the coalescing of overlapping variable names

There's an example [here](https://narwhals-dev.github.io/narwhals/why/) which shows
```
df_left = pd.DataFrame({'a': [1, 2, 3], 'b': [4, 5, 6]})
df_right = pd.DataFrame({'a': [1, 2, 3], 'c': [4, 5, 6]})
df_left.merge(df_right, left_on='b', right_on='c', how='left')
```

Could you include that as a test case please, so we can check that we get the same output across pandas and polars?

You might need to do some extra logic for the pandas one to make sure that the output names match

File: tests/test_common.py [hunk 3]
```diff
@@ -326,6 +338,38 @@ def test_join(df_raw: Any) -> None:
...
+        df_left_nw = nw.DataFrame(df_left)
+        df_right_nw = nw.DataFrame(df_right)
+        result_polars = df_left_nw.join(
+            df_right_nw, left_on="b", right_on="c", how="left"
+        )
+        result = nw.to_native(result_polars)
+        result = result.rename({"a": "a_left"}).to_pandas().to_dict(orient="list")
+
+    expected = {"a_left": [1, 2, 3], "b": [4, 5, 6], "a_right": [1, 2, 3], "c": [4, 5, 6]}
+    compare_dicts(result, expected)
```
Comment (MarcoGorelli) [inline tests/test_common.py:370]: the structure of the test should be more like:

- given
- when
- then

So:
- given: `df_pandas`, `df_pandas_right`, `df_polars`, `df_polars_right` (you've done this right)
- then: you do
  ```python
  df_left = nw.from_native(df_left)
  df_right = nw.from_native(df_right)
  result = df_left.join(df_right, left_on='b', right_on='c', how='left')
  ```
- then: you make `expected`, and check that `result` matches it

There shouldn't be any pandas-specific logic here. If there is, then it needs adjusting in  https://github.com/narwhals-dev/narwhals/blob/a22b600595559d9ba7950e3bc76badd4be76a548/narwhals/_pandas_like/dataframe.py#L249-L270

File: tests/test_common.py [hunk 4]
```diff
@@ -326,6 +338,45 @@ def test_join(df_raw: Any) -> None:
...
+        result_native = result.to_pandas()
+    else:
+        result_native = result
+
+    if isinstance(result_native, pd.DataFrame):
+        result_native = result_native.rename(columns={"a_x": "a", "a_y": "a_right"})
+        result_dict = result_native.to_dict(orient="list")
+    else:
+        result_native = result_native.rename({"a": "a", "a_right": "a_right"}).to_pandas()
+        result_dict = result_native.to_dict(orient="list")
```
Comment (MarcoGorelli) [inline tests/test_common.py:371]: as noted in https://github.com/narwhals-dev/narwhals/pull/312#discussion_r1648553370, we can't have logic in tests, this needs to go in `_pandas_like`

see https://github.com/narwhals-dev/narwhals/pull/312#discussion_r1648553370 for details

Comment (mfonekpo) [reply tests/test_common.py:371]: Alright Marco.

File: narwhals/_pandas_like/dataframe.py [hunk 5]
```diff
@@ -276,16 +276,28 @@ def join(
...
+        merged_df = self._dataframe.merge(
+            other._dataframe,
+            left_on=left_on,
+            right_on=right_on,
+            how=how,
+            suffixes=("", "_right"),
         )
 
+        start_col = "a"
+        end_col = "a_right"
```
Comment (MarcoGorelli) [inline narwhals/_pandas_like/dataframe.py:289]: what if someone has a dataframe where the column is called something different than `'a'`?

File: narwhals/_pandas_like/dataframe.py [hunk 6]
```diff
@@ -276,16 +276,28 @@ def join(
...
+        merged_df = self._dataframe.merge(
+            other._dataframe,
+            left_on=left_on,
+            right_on=right_on,
+            how=how,
+            suffixes=("", "_right"),
         )
 
+        start_col = merged_df.columns[0]
+        end_col = [col for col in merged_df.columns if col.endswith("_right")][0]
```
Comment (MarcoGorelli) [inline narwhals/_pandas_like/dataframe.py:289]: does this work even if there's multiple overlapping columns?

Comment (mfonekpo) [reply narwhals/_pandas_like/dataframe.py:289]: No, it doesn't. Let me try fix that

File: tests/frame/test_common.py [hunk 7]
```diff
@@ -312,6 +324,30 @@ def test_join(df_raw: Any) -> None:
     compare_dicts(result_native, expected)
 
 
+@pytest.mark.parametrize("constructor", [pl.DataFrame, pd.DataFrame])
+@pytest.mark.filterwarnings("ignore:the default coalesce behavior")
+def test_left_join(constructor: Any) -> None:
+    data_left = {"a": [1, 2, 3], "b": [4, 5, 6]}
+    data_right = {"a": [1, 2, 3], "c": [4, 5, 6]}
```
Comment (MarcoGorelli) [inline tests/frame/test_common.py:331]: can we add another test case where there's yet another overlapping column? like, add `'d': [1,4,2]` to both of them

Comment (mfonekpo) [reply tests/frame/test_common.py:331]: Alright.

File: tests/frame/test_common.py [hunk 8]
```diff
@@ -312,6 +324,42 @@ def test_join(df_raw: Any) -> None:
...
+
+
+@pytest.mark.parametrize("constructor", [pl.DataFrame, pd.DataFrame])
+@pytest.mark.filterwarnings("ignore: the defaultcoalesce behavior")
+def test_left_join_overlapping_column(constructor: Any) -> None:
+    data_left = {"a": [1, 2, 3], "b": [4, 5, 6], "d": [1, 4, 2]}
+    data_right = {"a": [1, 2, 3], "c": [4, 5, 6], "d": [1, 4, 2]}
+    df_left = nw.from_native(constructor(data_left))
+    df_right = nw.from_native(constructor(data_right))
+    result = df_left.join(df_right, left_on=["a", "b"], right_on=["a", "c"], how="left")
```
Comment (MarcoGorelli) [inline tests/frame/test_common.py:358]: can we just join on `'c'` here, so that there are two overlapping columns ('a' and 'd')?

**Comment by MarcoGorelli** [comment]
thanks for updating - could you also make sure to fetch and merge upstream/main, and resolve the merge conflicts?

File: narwhals/_pandas_like/dataframe.py [hunk 9]
```diff
@@ -345,6 +345,19 @@ def generate_unique_token(
             ),
         )
 
+        start_col = merged_df.columns[0]
+        end_col = [col for col in merged_df.columns if col.endswith("_right")][0]
```
Comment (MarcoGorelli) [inline narwhals/_pandas_like/dataframe.py:349]: this code you've added, it looks like it's outside the body of the function?

**Review by MarcoGorelli — COMMENTED** [review]
does it work to just do `merge` with `suffixes=("", "_right")`, and then drop any columns which only appear in `right_on` but not in `left_on`?

File: narwhals/_pandas_like/dataframe.py [hunk 10]
```diff
@@ -305,8 +305,8 @@ def join(
         other: Self,
         *,
         how: Literal["left", "inner", "outer", "cross", "anti", "semi"] = "inner",
-        left_on: str | list[str] | None = None,
-        right_on: str | list[str] | None = None,
+        left_on: str | list[str],
+        right_on: str | list[str],
```
[the remark this answers is not in the document]
Comment (MarcoGorelli) [reply narwhals/_pandas_like/dataframe.py:309]: right, thanks!

**Review by MarcoGorelli — APPROVED** [review]
thanks @mfonekpo , the tests you added were really helpful

sorry that this ended up being a lot harder than ancitipated, thanks anyway for your work here!

**Pull request merged**